core/os/midi_driver.cpp

```diff
@@ -43,76 +43,82 @@ void MIDIDriver::set_singleton() {
+		event.instantiate();
+		event->set_device(device_index);
 
-		case MIDIMessage::CONTROL_CHANGE:
-			if (length >= 2 + param_position) {
-				event->set_controller_number(data[param_position]);
-				event->set_controller_value(data[param_position + 1]);
+		if (length - index >= 1) {
+			if (data[index] >= 0xF0) {
+				// channel does not apply to system common messages
+				event->set_channel(0);
+				event->set_message(MIDIMessage(data[index]));
+				last_received_message = data[index];
+				index++;
+			} else if ((data[index] & 0x80) == 0x00) {
+				// running status, index is not changed because no new midi status byte, just additional data bytes
+				event->set_channel(last_received_message & 0xF);
+				event->set_message(MIDIMessage(last_received_message >> 4));
+			} else {
+				event->set_channel(data[index] & 0xF);
+				event->set_message(MIDIMessage(data[index] >> 4));
+				last_received_message = data[index];
+				index++;
 			}
-			break;
+		}
 
-		case MIDIMessage::NOTE_ON:
-		case MIDIMessage::NOTE_OFF:
-			if (length >= 2 + param_position) {
-				event->set_pitch(data[param_position]);
-				event->set_velocity(data[param_position + 1]);
-			}
-			break;
+		switch (event->get_message()) {
+			case MIDIMessage::AFTERTOUCH:
+				if (length >= 2 + index) {
+					event->set_pitch(data[index++]);
+					event->set_pressure(data[index++]);
+				}
+				break;
 
-		case MIDIMessage::PITCH_BEND:
-			if (length >= 2 + param_position) {
-				event->set_pitch((data[param_position + 1] << 7) | data[param_position]);
-			}
-			break;
+			case MIDIMessage::CONTROL_CHANGE:
+				if (length >= 2 + index) {
+					event->set_controller_number(data[index++]);
+					event->set_controller_value(data[index++]);
+				}
+				break;
 
-		case MIDIMessage::PROGRAM_CHANGE:
-			if (length >= 1 + param_position) {
-				event->set_instrument(data[param_position]);
-			}
-			break;
+			case MIDIMessage::NOTE_ON:
+			case MIDIMessage::NOTE_OFF:
+				if (length >= 2 + index) {
+					event->set_pitch(data[index++]);
+					event->set_velocity(data[index++]);
+				}
+				break;
 
-		case MIDIMessage::CHANNEL_PRESSURE:
-			if (length >= 1 + param_position) {
-				event->set_pressure(data[param_position]);
-			}
-			break;
-		default:
-			break;
-	}
+			case MIDIMessage::PITCH_BEND:
+				if (length >= 2 + index) {
+					event->set_pitch((data[index] << 7) | data[index]);
+					index++; // clearer than trying to increment on line above given the or operator
+				}
+				break;
+
+			case MIDIMessage::PROGRAM_CHANGE:
+				if (length >= 1 + index) {
+					event->set_instrument(data[index++]);
+				}
+				break;
 
-	Input *id = Input::get_singleton();
-	id->parse_input_event(event);
+			case MIDIMessage::CHANNEL_PRESSURE:
+				if (length >= 1 + index) {
+					event->set_pressure(data[index++]);
+				}
+				break;
+			default:
+				break;
+		}
+
+		Input *id = Input::get_singleton();
+		id->parse_input_event(event); // new InputEventMidi for each message read from *data
```

Comment:
```suggestion
		Input::get_singleton()->parse_input_event(event);
```
No need to fetch as a variable for one line, comment unnecessary